oauth_dropins/meetup.py

```diff
@@ -0,0 +1,162 @@
+from .models import BaseAuth
+from .webutil import appengine_config, appengine_info, util
+from .webutil.util import json_dumps, json_loads
+
+if appengine_info.DEBUG:
+    MEETUP_CLIENT_ID = util.read('meetup_client_id_local')
+    MEETUP_CLIENT_SECRET = util.read('meetup_client_secret_local')
+else:
+    MEETUP_CLIENT_ID = util.read('meetup_client_id')
+    MEETUP_CLIENT_SECRET = util.read('meetup_client_secret')
+
+GET_AUTH_CODE_URL = '&'.join((
+    'https://secure.meetup.com/oauth2/authorize'
+    'client_id=%{client_id}s',
+    'response_type=code',
+    'redirect_uri=%{redirect_uri}s',
+    'state=%(state)s',
+    'scope=rsvp'
+    ))
+
+GET_ACCESS_TOKEN_URL = '&'.join((
+    'https://secure.meetup.com/oauth2/access'
+    'grant_type=authorization_code',
+    'code=%(code)s',
+    'client_id=%(client_id)s',
+    'client_secret=%(client_secret)s',
+    'redirect_uri=%(redirect_uri)s',
+    ))
+
+GET_USER_INFO_URL = 'https://api.meetup.com/members/self/'
+
+
+class MeetupAuth(BaseAuth):
+    """An authenticated Meetup.com user.
+
+    Provides methods that return information about this user and make
+    OAuth-signed requests to Meetup's HTTP-based APIs. Stores OAuth credentials
+    in the datastore. See models.BaseAuth for usage details.
+
+    Implements urlopen() but not http() or api().
+    """
+    access_token_str = ndb.StringProperty(required=True)
+
+    def site_name(self):
+        return 'Meetup.com'
+
+    def user_display_name(self):
+        """Returns the Meetup.com user id.
+        """
+        return self.key.string_id()
+
+
+def access_token(self):
+    """Returns the OAuth access token string.
+    """
+    return self.access_token_str
+
+
+def urlopen(self, url, **kwargs):
+    """Wraps urllib2.urlopen() and adds OAuth credentials to the request.
+    """
+    headers = {'Authorization': 'Bearer %s' % self.access_token_str}
+    try:
+        return util.urlopen(urllib2.Request(url, headers=headers), **kwargs)
+    except BaseException as e:
+        util.interpret_http_exception(e)
+        raise
+
+
+class MeetupCsrf(ndb.Model):
+    """Stores a CSRF token for the Meetup.com OAuth2 flow."""
+    token = ndb.StringProperty(required=False)
+    state = ndb.TextProperty(required=False)
+
+
+class StartHandler(handlers.StartHandler):
+    """Starts Meetup.com auth. Requests an auth code and expects a redirect back.
+    """
+    def redirect_url(self, state=None):
+        assert (MEETUP_CLIENT_ID and MEETUP_CLIENT_SECRET), (
+                "Please fill in the meetup_client_id and meetup_client_secret files in "
+                "your app's root directory.")
+
+        csrf_key = MeetupCsrf(state=state).put()
+        return GET_AUTH_CODE_URL % {
+                'client_id': MEETUP_CLIENT_ID,
+                'redirect_uri': urllib.quote_plus(self.to_url(state=state)),
+                'state': '%s|%s' % (state, csrf_key.id()),
+                }
+
+
+class CallbackHandler(handlers.CallbackHandler):
```

Comment:
you'll want to override the `button_html()` method here to set a non-white background color. (assuming the logo pngs are transparent background?)

Reply:
It turns out the image has a white background so this shouldn't be needed :+1:

Reply:
oh, sorry if i was unclear. all of the buttons (ie logos) do need non-white backgrounds. you're welcome to do that either in the image itself, or in CSS by overriding `button_html()` with a transparent background image.

Reply:
Sorry, my misinterpretation - I've corrected that now